en/lessons/libraries/mox.md

```diff
@@ -0,0 +1,196 @@
+
+Remember: as a pre-requisite, our code must look to the application config to get its configured module:
+
+```elixir
+def get_username(username) do
+  http_client().get("https://elixirschool.com/users/#{username}")
+end
+
+defp http_client do
+  Application.get_env(:my_app, :http_client)
+end
+```
+
+Then you may include `mox` in your dependencies:
+
+```elixir
+# mix.exs
+defp deps do
+  [
+    # ...
+    {:mox, "~> 0.5.2", only: [:test], runtime: false}
+  ]
+end
+```
+
+Install it with `mix deps.get`.
+
+Next, modify your `test_helper.exs` so it does 3 things:
+
+1. it must start the Mox app
+2. it must define one or more mocks
+3. it must set the application config with the mock
+
+```elixir
+# test_helper.exs
+ExUnit.start(exclude: [:skip])
+
+# 1. Start the mox app
+Mox.Server.start_link([])
+
+# 2. define dynamic mocks
+Mox.defmock(HTTPoison.BaseMock, for: HTTPoison.Base)
+# ... etc...
+
+# 3. Override the config settings (similar to adding these to `config/test.exs`)
+Application.put_env(:my_app, :http_client, HTTPoison.BaseMock)
+# ... etc...
+```
+
+A couple important things to note about `Mox.defmock`: the left-side name is arbitrary. Module names in Elixir are just atoms -- you don't have to create a module anywhere, all you are doing is "reserving" a name for the mock module. Behind the scenes, Mox will create a module with this name on the fly inside the BEAM.
+
+The second tricky thing is that the module referenced by `for:` _must_ be a behaviour: it _must_ define callbacks. Mox uses introspection on this module and you can only define mock functions when a `@callback` has been defined.  This is how Mox enforces a contract.  Sometimes it can be difficult to find the behaviour module: `HTTPoison` for example, relies on `HTTPoison.Base`, but you might not know that unless you look through its source code. If you are trying to create a mock for a 3rd-party package, you may discover that no behaviour exists! In those cases you may need to define your own behaviour and callbacks to satisfy the need for a contract.
+
+This brings up an important point: you may want to use a layer of abstraction (a.k.a. [indirection](https://en.wikipedia.org/wiki/Indirection)) so your application doesn't depend on a third party package _directly_, but instead you would use your own module which in turn uses the package. It is important for a well-crafted application to define the proper "boundaries", but the mechanics of mocks do not change, so don't let that trip you up.
+
+Finally, in your test modules, you can put your mocks to use by importing `Mox` and calling its `:verify_on_exit!` function. Then you are free to define return values on your mock modules using one or more calls to the `expect` function:
+
+```elixir
+defmodule MyAppTest do
+  use ExUnit.Case
+  # 1. Import Mox
+  import Mox
+  # 2. setup fixtures
+  setup :verify_on_exit!
+
+  test ":ok on 200" do
+    HTTPoison.BaseMock
+    |> expect(:get, fn _ -> {:ok, "What a guy!"} end)
```

Comment:
We don't use pipelines with only a single function call in our other examples so let's stick with that :+1: 

```suggestion
    expect(HTTPoison.BaseMock, :get, fn _ -> {:ok, "What a guy!"} end)
```